src/main/java/com/google/devtools/build/lib/rules/proto/ProtoCommon.java

```diff
@@ -491,57 +466,45 @@ private static boolean hasValidMigrationTag(RuleContext ruleContext) {
    * ruleContext}.
    */
   public static ProtoInfo createProtoInfo(
-      RuleContext ruleContext, boolean generatedProtosInVirtualImports)
-      throws InterruptedException {
-    checkSourceFilesAreInSamePackage(ruleContext);
-    ImmutableList<Artifact> directProtoSources =
-        ruleContext.getPrerequisiteArtifacts("srcs", Mode.TARGET).list();
-    Library library =
-        createLibraryWithVirtualSourceRootMaybe(
-            ruleContext, directProtoSources, generatedProtosInVirtualImports);
-    if (ruleContext.hasErrors()) {
-      return null;
-    }
-
-    if (library == null) {
-      library = createLibraryWithoutVirtualSourceRoot(ruleContext, directProtoSources);
-    }
+      ImmutableList<Artifact> originalDirectProtoSources,
```

Comment:
What's code is coming up? There are a few issues here:

* This makes the interface of `createProtoInfo` quite a bit more verbose
* It relies on the caller correctly computing `protoSourceRoot` and `sourceImportPairs`

Reply:
Yes, it is more verbose but it is also more accurate.  The ProtoInfo class should only have information related to compiling protos.  Currently we're pulling the protoSourceRoot and sourceImportPairs information out of the ruleContext which lets us hide a lot of the interface in the ruleContext.    I think it makes more sense to calculate these variables outside of the createProtoInfo function and pass them in.

Reply:
I think that's reasonable, but you'll need to find a way to prevent people passing in inconsistent data (e.g. not every item in `directProtoSources` having an entry in `sourceImportPathPairs` or `sourceImportPathPairs` inconsistent with `protoSourceRoot`.

One possible way to go about this is to have two methods: one to create a `Library` that's guaranteed to be consistent and has access to `ctx` and `createProtoInfo()` like you propose here. I can't think up a way to avoid accessing `ctx` for during the creation of the `Library` because, well, it at least needs to know in which repository it is.

Reply:
Mind if I add verification in a second PR?  Is adding a TODO: tag ok for now?

Reply:
Mind doing it in this change? I have become very wary of "temporary" lifting of restrictions because even with the best intent, they have the tendency to become permanent. And it's not more work to do it in one change.

What do you think about the idea of wrapping the creation of `Library` in a Starlark call instead? That way, you wouldn't need verification because you'd be guaranteed to have it always correct. Something like:

```
lib = proto_common.create_library(ctx.actions, [proto files], strip_import_prefix, import_prefix)
info = proto_common.create_info(lib, [ProtoInfo instances])
```

It would then create symlink actions if needed and if `strip_import_prefix` is an exec path then `create_library` doesn't need `ctx` only `ctx.actions`, thereby limiting its power.

Reply:
Having two functions that are only useful together doesn't sound like a nice API (to me). Depending on what the use-case for the API is, we might have more success by inventing `proto_common.compile` (or `.create_descriptor_set`, naming is hard) which does the same thing in one step.

```starlark
info = proto_common.compile(
    actions = ctx.actions,
    srcs = [sequence of `.proto` files],
    deps = [sequence of ProtoInfo],
    exports = [sequence of ProtoInfo],
    import_prefix = None,
    # maybe just exec path, maybe `absolute -> exec path`, `relative -> relative to a file's owner's package`?
    strip_import_prefix = None,
)
```

Reply:
Ok, removed the second method from this PR instead.  I think it will be better to discuss it in the follow up PR with the new Starlark method.

Reply:
@lberki  do you mind taking another look at this?  I removed the second createProtoInfo method so this PR is only refactoring existing methods.